Move YAML strings to YAML arrays for CORS configuration

containscafeine opened the pull request:
This commit makes defining CORS configuration in ambassador more
readable by moving comma separated strings, which envoy supports,
to YAML arrays.

The previous format looks like -
```yaml
cors:
  origins: http://foo.example,http://bar.example
  methods: POST, GET, OPTIONS
  headers: Content-Type,Accept,Authorization
  credentials: true
  exposed_headers: X-Custom-Header
  max_age: "86400"
```

The new format looks like -
```yaml
cors:
  origins:
  - http://foo.example
  - http://bar.example
  methods:
  - POST
  - GET
  - OPTIONS
  headers:
  - Content-Type
  - Accept
  - Authorization
  credentials: true
  exposed_headers:
  - X-Custom-Header
  max_age: "86400"
```

Fix #466

containscafeine commented:
@kflynn yet to fix docs and add comments, but is the approach fine?

richarddli commented:
Does this change require a move to YAML arrays? We want to make sure we don't break backwards compatibility.

containscafeine commented:
Oops, yes this is backward incompatible :man_facepalming:

alexgervais commented:
My 2 cents, but the reason these were strings is because the values map to HTTP headers, which are themselves CSV values:
```
Access-Control-Allow-Methods: POST, GET
Access-Control-Allow-Headers: X-PINGOTHER, Content-Type
```

I also find it strange for `Access-Control-Allow-Origin` to be an array containing a single `*` value.

kflynn commented:
In other places (e.g. the `ambassador_id` element) we allow both strings and arrays, so any of these would be fine:

```
cors:
  methods: POST, GET

cors:
  methods:
  - POST
  - GET

cors:
  methods: [ "POST", "GET" ]
```

I don't want to require the array syntax. I'm fine with supporting it, though.

plombardi89 commented:
I don't think I would find `["*"]` too weird in @alexgervais case. It is just configuration after all. Since the real request here was about improving readability I'm wondering if we have one other option: Could folks just use a multi line YAML string? Does Envoy know how to handle newlines?

I think I still prefer the structured format in general though. IMO I would not long term support the non-array format, but that's a different issue.

containscafeine commented:
Okay, I've made the changes to support both, the older and the newer syntax. Updated and added tests and docs to reflect this. Not particularly happy about the docs update, PTAL.

plombardi89 commented:
Is there a particular concern you have wrt the doc update?

containscafeine commented:
@plombardi89 it just does not look pretty, that's all!

kflynn commented:
I think the docs are OK.

kflynn commented on ambassador/ambassador/mapping.py:
```diff
@@ -122,6 +122,46 @@ def get(self, key, *args):
 
+    @staticmethod
+    def add_to_dict_if_key_exists(key1, dict1, key2, dict2):
+        """If dict1[key1] exists, and
```
This should probably be something like `cors_save_element`. That's semantically narrow enough that it's OK for it to modify `self.cors` instead of having to have a dict passed in. 

```python
def cors_save_element(self, cors_key, params, param_key):
    ...
    self.cors[cors_key] = params[param_key] # or whatever
    ...
```

and, later, 

```python
self.cors_save_element('credentials', route_cors, 'allow_credentials')
```

containscafeine replied:
makes sense, fixed

kflynn commented on ambassador/ambassador/mapping.py:
```diff
@@ -122,6 +122,46 @@ def get(self, key, *args):
+        """Generates envoy's cors configuration from ambassador's cors configuration
+
+        :param cors: dict that contains ambassador's representation of cors
+        :param route_cors: envoy's route's cors configuration dict
```
Same here -- that's just `self.cors`, no?

kflynn commented on ambassador/ambassador/mapping.py:
```diff
@@ -166,16 +206,9 @@ def new_route(self, svc, cluster_name):
-            }.items() if value}
+        if cors is not None:
+            route['cors'] = dict()
+            self.generate_route_cors(cors, route['cors'])
```
This should probably look like

```python
route['cors`] = self.generate_route_cors(cors)
```

instead of using the `route['cors']` dict as an output parameter.

containscafeine replied:
fixed

kflynn commented:
Some comments inline. Sorry for flipflopping on some review comments!

containscafeine commented on ambassador/ambassador/mapping.py:
```diff
@@ -122,6 +122,46 @@ def get(self, key, *args):
+                route_cors['allow_origin'] = cors.get('origins')
+            elif type(cors.get('origins')) is str:
+                route_cors['allow_origin'] = cors.get('origins').split(',')
+
```
I've added -
```python
            else:
                print("invalid cors configuration supplied - {}".format(cors.get('origins')))
                return 
```
not sure what the idiomatic way to do this is.

containscafeine commented:
@kflynn np, the more reviews the merrier. Made the changes, PTAL.

kflynn commented on ambassador/ambassador/mapping.py:
```diff
@@ -122,6 +122,53 @@ def get(self, key, *args):
+        :param route: envoy's cors configuration
+        """
+        cors = self.get('cors')
+        if cors.get(cors_key) is not None:
```
Won't this raise exceptions if `cors_key` isn't present in `cors`? (This applies to many places here... sorry I missed it last time!)

containscafeine replied:
@kflynn will it? Just tried -
```python
>>> a={"a":1,"b":2}
>>> if a.get("c") is not None:
...     print("hi")
... 
```
and it didn't raise an exception even though "c" is not present in dict a.

kflynn replied:
...just rechecked the docs and I don't know WTF I was thinking. :) :) You're right.

kflynn commented on ambassador/ambassador/mapping.py:
```diff
@@ -122,6 +122,53 @@ def get(self, key, *args):
+        # cors['origins'] cannot be treated like other keys, because if it's a
+        # list, then it remains as is, but if it's a string, then it's
+        # converted to a list
+        if cors.get('origins') is not None:
```
This block will be way simplified by using a temporary variable for `cors.get('origins')` -- it won't make a difference in terms of speed, of course, but it'll make a difference in terms of ease of understanding for the next guy.

containscafeine replied:
fixed

containscafeine commented:
:tada:
